### ISSUE TITLE
riscv64: Miri UI tests fail on riscv64gc-unknown-linux-gnu

### ISSUE THREAD

**Issue by yuzibo** [rust-lang/rust#155100]
<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

The issue first occurred back [03/17](https://community-ci.openruyi.cn/rvci/job/RustCI/job/RustCheck/826/) from the records.

```rust
jenkins@a994a9690b9f:~/rust$ ./x test --stage 2 src/tools/miri                                                                    
Building bootstrap                                                                                                                
    Finished `dev` profile [unoptimized] target(s) in 2.71s                                                                       
WARNING: The `change-id` is missing in the `bootstrap.toml`. This means that you will not be able to track the major changes made to the bootstrap configurations.
NOTE: to silence this warning, add `change-id = 154508` or `change-id = "ignore"` at the top of `bootstrap.toml`                  
/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/ci-llvm/bin/llvm-strip does not exist; skipping copy                         
Building stage1 compiler artifacts (stage0 -> stage1, riscv64gc-unknown-linux-gnu)                                                
    Finished `release` profile [optimized] target(s) in 3.80s                                                                     
Creating a sysroot for stage1 compiler (use `rustup toolchain link 'name' build/host/stage1`)                                                                                                                                                                       
/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/ci-llvm/bin/llvm-strip does not exist; skipping copy                                                                                                                                                           Building stage1 library artifacts (stage1 -> stage1, riscv64gc-unknown-linux-gnu)                                                                                                                                                                                   
    Finished `dist` profile [optimized] target(s) in 0.50s                                                                                                                                                                                                          
Building stage2 compiler artifacts (stage1 -> stage2, riscv64gc-unknown-linux-gnu)                                                                                                                                                                                  
    Finished `release` profile [optimized] target(s) in 4.23s                                                                                                                                                                                                       
Creating a sysroot for stage2 compiler (use `rustup toolchain link 'name' build/host/stage2`)                                     
Building stage2 miri (stage1 -> stage2, riscv64gc-unknown-linux-gnu)                                                                                                                                                                                                
...
     Running tests/ui.rs (build/riscv64gc-unknown-linux-gnu/stage2-tools/riscv64gc-unknown-linux-gnu/release/deps/ui-7f737c2d72867e53)                                                                                                                              
## Running ui tests in tests/pass for riscv64gc-unknown-linux-gnu                                                                                                                                                                                                   
   Compiler: "MIRI_ENV_VAR_TEST"="0" "MIRI_TEMP"="/tmp/miri-uitest-mjTy1H" "RUST_BACKTRACE"="1" /home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2-tools/riscv64gc-unknown-linux-gnu/release/miri "--error-format=json" "--sysroot=/home/jenkins/rust/build
/riscv64gc-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" OUT_DIR       

FAILED TEST: tests/pass/atomic.rs (revision `stack`)                                                                              
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-mjTy1H" RUST_BACKTRACE="1" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2-tools/riscv64gc-unknown-linux-gnu/release/miri" "--error-format=json" "--sysroot=/home/jenkins/rust/build/riscv64
gc-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2-tools/riscv64gc-unknown-linux-gnu/tmp/miri_ui/0/tests/pass" "tests/pass/atomic.rs" "-Zmiri-
strict-provenance" "--edition" "2021" "--cfg=stack" "-Cextra-filename=stack"                                                      
                                                                                                                                  
error: test got exit status: 101, but expected 0                                                                                  
 = note: the compiler panicked                                                                                                    
                                                                                                                                  
error: no output was expected                                                                                                     
Execute `./miri test --bless` to update `tests/pass/atomic.stack.stderr` to the actual output                                     
+++ <stderr output>                                              
                                                                                                                                  
thread 'main' (1000) panicked at tests/pass/atomic.rs:198:5:                                                                      
compare_exchange_weak succeeded 100 tries in a row                                                                                
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace                                                     
note: in Miri, you may have to set `MIRIFLAGS=-Zmiri-env-forward=RUST_BACKTRACE` for the environment variable to have an effect   
                                                                                                                                  
                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                    
FAILED TEST: tests/pass/atomic.rs (revision `tree`)                                                                               
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-mjTy1H" RUST_BACKTRACE="1" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2-tools/riscv64gc-unknown-linux-gnu/release/miri" "--error-format=json" "--sysroot=/home/jenkins/rust/build/riscv64
gc-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2-tools/riscv64gc-unknown-linux-gnu/tmp/miri_ui/0/tests/pass" "tests/pass/atomic.rs" "-Zmiri-
strict-provenance" "-Zmiri-tree-borrows" "--edition" "2021" "--cfg=tree" "-Cextra-filename=tree"                                  
                                                                                                                                  
error: test got exit status: 101, but expected 0                                                                                  
 = note: the compiler panicked                                                                                                    
                                                                 
error: no output was expected                                                                                                     
Execute `./miri test --bless` to update `tests/pass/atomic.tree.stderr` to the actual output                                      
+++ <stderr output>                                                                                                               
                                                                                                                                                                                                                                                                    
thread 'main' (1000) panicked at tests/pass/atomic.rs:198:5:                                                                      
compare_exchange_weak succeeded 100 tries in a row                                                                                
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace                                                     
note: in Miri, you may have to set `MIRIFLAGS=-Zmiri-env-forward=RUST_BACKTRACE` for the environment variable to have an effect   
                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
FAILED TEST: tests/pass/shims/socket.rs                                                                                                                                                                                                                             
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-mjTy1H" RUST_BACKTRACE="1" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2-tools/riscv64gc-unknown-linux-gnu/release/miri" "--error-format=json" "--sysroot=/home/jenkins/rust/build/riscv64
gc-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2-tools/riscv64gc-unknown-linux-gnu/tmp/miri_ui/0/tests/pass/shims" "tests/pass/shims/socket.
rs" "-Zmiri-disable-isolation" "--edition" "2021"                                                                                 
                                                                                                                                                                                                                                                                    
error: test got exit status: 101, but expected 0                                                                                                                                                                                                                    
 = note: the compiler panicked                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                    
error: no output was expected                                                                                                                                                                                                                                       
Execute `./miri test --bless` to update `tests/pass/shims/socket.stderr` to the actual output                                                                                                                                                                       
+++ <stderr output>                                                                                                               
                                                                                                                                                                                                                                                                    
thread 'main' (7683) panicked at tests/pass/shims/socket.rs:24:55:                                                                                                                                                                                                  
called `Result::unwrap()` on an `Err` value: Os { code: 99, kind: AddrNotAvailable, message: "address not available" }            

called `Result::unwrap()` on an `Err` value: Os { code: 99, kind: AddrNotAvailable, message: "address not available" }
stack backtrace:
   0: std::panicking::panic_handler
             at /home/jenkins/rust/library/std/src/panicking.rs:689:5
   1: std::rt::panic_fmt
             at /home/jenkins/rust/library/core/src/panicking.rs:80:14
   2: std::result::unwrap_failed
             at /home/jenkins/rust/library/core/src/result.rs:1867:5
   3: std::result::Result::<std::net::TcpListener, std::io::Error>::unwrap
             at /home/jenkins/rust/library/core/src/result.rs:1233:23
   4: test_create_ipv6_listener
             at tests/pass/shims/socket.rs:24:26
   5: main
             at tests/pass/shims/socket.rs:12:5
   6: <fn() as std::ops::FnOnce<()>>::call_once - shim(fn())
             at /home/jenkins/rust/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


FAILURES:
    tests/pass/atomic.rs (revision stack)
    tests/pass/atomic.rs (revision tree)
    tests/pass/shims/socket.rs

test result: FAIL. 3 failed; 372 passed; 29 ignored

Error: 
   0: ui tests in tests/pass for riscv64gc-unknown-linux-gnu failed
   1: tests failed

Location:
   /home/jenkins/.cargo/registry/src/community-ci.openruyi.cn-563c67b1241fa9eb/ui_test-0.30.3/src/lib.rs:365

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`

Caused by:
  process didn't exit successfully: `/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2-tools/riscv64gc-unknown-linux-gnu/release/deps/ui-7f737c2d72867e53 --quiet` (exit status: 1)
Command `/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage0/bin/cargo test -Zwarnings --target riscv64gc-unknown-linux-gnu -Zbinary-dep-depinfo -j 64 -Zroot-dir=/home/jenkins/rust --profile=release --manifest-path /home/jenkins/rust/src/tools/miri/Cargo.toml -- --quiet [workdir=/home/jenkins/rust]` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:191:21
Executed at: src/bootstrap/src/core/build_steps/test.rs:740:19

Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:16:49

```

**Comment by yuzibo** [comment]
Just for socket.rs:
This seems to be due to missing IPv6 support in the environment (AddrNotAvailable).

**Comment by saethlin** [comment]
The atomic behavior is because riscv gets EMULATE_ATOMIC_BOOL:
https://github.com/rust-lang/rust/blob/bf4fbfb7a18d74e7cd8eef93af7329c58fbb5344/library/core/src/sync/atomic.rs#L375-L388

Since that is papering over a codegen problem, it should be disabled for `cfg(miri)`.

(It would be even better if LLVM didn't need this assistance to do what we want)

**Comment by RalfJung** [comment]
> Since that is papering over a codegen problem, it should be disabled for cfg(miri).

`cfg(miri)` in the standard library should be the very last resort. We want to test the code people actually run.
Better to disable that part of the test on riscv than to change the standard library.

**Comment by SpriteOvO** [comment]
@yuzibo Hi, I opened 2 PRs (rust-lang/miri#4971 and rust-lang/miri#4972) to fix the test errors for RISC-V, could you test if they work in your environment? I have tested them in QEMU RISC-V, and they look fine to me.

**Comment by yuzibo** [comment]
> [@yuzibo](https://github.com/yuzibo) Hi, I opened 2 PRs ([rust-lang/miri#4971](https://github.com/rust-lang/miri/pull/4971) and [rust-lang/miri#4972](https://github.com/rust-lang/miri/pull/4972)) to fix the test errors for RISC-V, could you test if they work in your environment? I have tested them in QEMU RISC-V, and they look fine to me.

@SpriteOvO Thank you for fixing this! 

I believe the `IPV6` [issue](https://github.com/rust-lang/miri/pull/4972/changes/6c2d862d2ee5580c7a4a83d6cc2fa8d4b454614a#r3115687395) should be from our runner. I will test them on our runner next days again.
